src/Internal/BalanceTx/Error.purs

```diff
@@ -89,6 +90,37 @@ instance Show BalanceTxError where
     "ExUnitsEvaluationFailed: " <> printTxEvaluationFailure tx failure
   show e = genericShow e
 
+explainBalanceTxError :: BalanceTxError -> String
+explainBalanceTxError = case _ of
+  BalanceInsufficientError expected actual ctx ->
+    "Insufficient balance. Expected: " <> show expected
+      <> ", actual: "
+      <> show actual
+      <> ". Context: "
+      <> show ctx
+  CouldNotConvertScriptOutputToTxInput ->
+    "Could not convert script output to transaction input."
+  CouldNotGetChangeAddress ->
+    "Could not get change address."
+  CouldNotGetCollateral -> "Could not get collateral."
+  CouldNotGetUtxos -> "Could not get UTxOs."
```

Comment:
Depends. The source of this error can have two possibilities:

```haskell
    utxos <- liftEitherContract do
      case mbSrcAddrs of
        -- Use wallet UTxOs.
        Nothing -> do
          whenM
            ( asks $ _.synchronizationParams
                >>> _.syncBackendWithWallet
                >>> _.beforeBalancing
            )
            syncBackendWithWallet
          note couldNotGetUtxos <$> do
            Wallet.getWalletUtxos
        -- Use UTxOs from source addresses
        Just srcAddrs -> do
          queryHandle <- getQueryHandle
          -- Even though some of the addresses may be controlled by the wallet,
          -- we can't query the wallet for available UTxOs, because there's no
          -- way to tell it to return UTxOs only from specific subset of the
          -- addresses controlled by a CIP-30 wallet.
          -- `utxosAt` calls are expensive when there are a lot of addresses to
          -- check.
          parTraverse (queryHandle.utxosAt >>> liftAff >>> map hush) srcAddrs
            <#> traverse (note CouldNotGetUtxos)
              >>> map (foldr Map.union Map.empty) -- merge all utxos into one map
```

Should we add additional information to specify whether we couldn't find wallet UTxOs or source address UTxOs?

Reply:
so, "from the wallet or from the specified source addresses"